### PR TITLE
[SPARK-9836] [ML] Provide R-like summary statistics for OLS via normal equation solver

### PR DESCRIPTION
**Pull request by yanboliang**
https://issues.apache.org/jira/browse/SPARK-9836

### CODE REVIEW

**Comment by yanboliang** [comment]
In the current implementation we provide ```Std. Error``` for ```coefficients``` excepts ```intercept```, because that we use optimized method to calculate ```intercept```. If we want to calculate ```Std. Error``` for ```intercept```, we need to concat ```aBar``` array with ```aaBar.values``` like
```scala
val newAtA = Array.concat(aaBar.values, summary.aBar.toArray, Array(1.0))
val newAtB = Array.concat(abBar.values, Array(bBar))

val xWithIntercept = CholeskyDecomposition.solve(newAtA, newAtB)
val newAtAi = CholeskyDecomposition.inverse(newAtA, summary.k)
```
I'm afraid that it will cause performance degradation, so I propose output ```Std. Error``` only for ```coefficients``. May be here we should discuss, or figure out better way to output ```Std. Error``` for ```intercept``. @mengxr

**Comment by yanboliang** [comment]
In the current implementation we provide ```Std. Error``` for ```coefficients``` excepts ```intercept```, because that we use optimized method to calculate ```intercept```. If we want to calculate ```Std. Error``` for ```intercept```, we need to concat ```aBar``` array with ```aaBar.values``` like
```scala
val newAtA = Array.concat(aaBar.values, summary.aBar.toArray, Array(1.0))
val newAtB = Array.concat(abBar.values, Array(bBar))

val xWithIntercept = CholeskyDecomposition.solve(newAtA, newAtB)
val newAtAi = CholeskyDecomposition.inverse(newAtA, summary.k)
```
I'm afraid that it will cause performance degradation, so I propose output ```Std. Error``` only for ```coefficients``. May be here we should discuss, or figure out better way to output ```Std. Error``` for ```intercept``. @mengxr

**Comment by yanboliang** [comment]
In the current implementation we provide ```Std. Error``` for ```coefficients``` excepts ```intercept```, because that we use optimized method to calculate ```intercept```. If we want to calculate ```Std. Error``` for ```intercept```, we need to concat ```aBar``` array with ```aaBar.values``` like
```scala
val newAtA = Array.concat(aaBar.values, summary.aBar.toArray, Array(1.0))
val newAtB = Array.concat(abBar.values, Array(bBar))

val xWithIntercept = CholeskyDecomposition.solve(newAtA, newAtB)
val newAtAi = CholeskyDecomposition.inverse(newAtA, summary.k)
```
I'm afraid that it will cause performance degradation, so I propose output ```Std. Error``` only for ```coefficients``. May be here we should discuss, or figure out better way to output ```Std. Error``` for ```intercept``. @mengxr

**Comment by yanboliang** [comment]
In the current implementation we provide ```Std. Error``` for ```coefficients``` excepts ```intercept```, because that we use optimized method to calculate ```intercept```. If we want to calculate ```Std. Error``` for ```intercept```, we need to concat ```aBar``` array with ```aaBar.values``` like
```scala
val newAtA = Array.concat(aaBar.values, summary.aBar.toArray, Array(1.0))
val newAtB = Array.concat(abBar.values, Array(bBar))

val xWithIntercept = CholeskyDecomposition.solve(newAtA, newAtB)
val newAtAi = CholeskyDecomposition.inverse(newAtA, summary.k)
```
I'm afraid that it will cause performance degradation, so I propose output ```Std. Error``` only for ```coefficients``. May be here we should discuss, or figure out better way to output ```Std. Error``` for ```intercept``. @mengxr

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 1]
```diff
@@ -471,6 +484,59 @@ class LinearRegressionSummary private[regression] (
...
+    numInstances - model.weights.size
+  }
+
+  lazy val devianceResiduals: Array[Double] = {
+    val weighted = if (model.getWeightCol.isEmpty) lit(1.0) else sqrt(col(model.getWeightCol))
+    val dr = predictions.select(col(model.getLabelCol).minus(col(model.getPredictionCol))
+      .multiply(weighted).as("weightedResiduals"))
+      .select(min(col("weightedResiduals")).as("min"), max(col("weightedResiduals")).as("max"))
+      .take(1)(0)
+    Array(dr.getDouble(0), dr.getDouble(1))
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: DataFrame currently does not provide interface to calculate percentile (only Hive UDAF), so here we only provide max and min value of deviance residuals. [SPARK-9299](https://issues.apache.org/jira/browse/SPARK-9299) works on providing ```percentile``` and ```percentile_approx``` aggregate functions, after it was resolved we can provide deviance residuals of quantile (0.25, 0.5, 0.75).

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala [hunk 2]
```diff
@@ -26,10 +26,12 @@ import org.apache.spark.rdd.RDD
  * Model fitted by [[WeightedLeastSquares]].
  * @param coefficients model coefficients
  * @param intercept model intercept
+ * @param diag diagonal of matrix (A^T * W * A)^-1
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: `diag` is not a descriptive name, `diagInvAtWA?

File: mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala [hunk 3]
```diff
@@ -109,6 +113,9 @@ private[ml] class WeightedLeastSquares(
 
     val x = new DenseVector(CholeskyDecomposition.solve(aaBar.values, abBar.values))
 
+    val aaInv = CholeskyDecomposition.inverse(aaBar.values, k)
+    val diag = new DenseVector((1 to k).map{ i => aaInv(i + (i - 1) * i / 2 - 1) / wSum }.toArray)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala]: Need an inline comment to explain the index mapping. It is sufficient to just mention that `aaInv` is a packed upper triangular matrix.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 4]
```diff
@@ -471,6 +484,59 @@ class LinearRegressionSummary private[regression] (
     predictions.select(t(col(predictionCol), col(labelCol)).as("residuals"))
   }
 
+  lazy val numInstances: Long = predictions.count()
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: missing doc (please also update other public/private methods)

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 5]
```diff
@@ -471,6 +484,59 @@ class LinearRegressionSummary private[regression] (
...
+
+  lazy val dfe = if (model.getFitIntercept) {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Keep this one private, or a more descriptive name? We need explicit types for public/private members.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 6]
```diff
@@ -471,6 +484,59 @@ class LinearRegressionSummary private[regression] (
...
+  } else {
+    numInstances - model.weights.size
+  }
+
+  lazy val devianceResiduals: Array[Double] = {
+    val weighted = if (model.getWeightCol.isEmpty) lit(1.0) else sqrt(col(model.getWeightCol))
+    val dr = predictions.select(col(model.getLabelCol).minus(col(model.getPredictionCol))
+      .multiply(weighted).as("weightedResiduals"))
+      .select(min(col("weightedResiduals")).as("min"), max(col("weightedResiduals")).as("max"))
+      .take(1)(0)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: `.take(1)(0`) -> `.first()`

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 7]
```diff
@@ -471,6 +484,59 @@ class LinearRegressionSummary private[regression] (
...
+    Array(dr.getDouble(0), dr.getDouble(1))
+  }
+
+  lazy val seCoef: Array[Double] = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: `coefficientStandardErrors`? It is hard to guess what `seCoef` means. In the doc, we should say "intercept" is not supported.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 8]
```diff
@@ -471,6 +484,59 @@ class LinearRegressionSummary private[regression] (
...
+          math.pow(label - pred, 2.0) * weight }
+        predictions.select(t(col(model.getPredictionCol), col(model.getLabelCol),
+          col(model.getWeightCol)).as("wse")).agg(sum(col("wse"))).take(1)(0).getDouble(0)
+      }
+      val sigma2 = rss / dfe
+      diag.map(_ * sigma2).map(math.sqrt(_))
+    }
+  }
+
+  lazy val tVals: Array[Double] = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: `tValues`

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 9]
```diff
@@ -471,6 +484,59 @@ class LinearRegressionSummary private[regression] (
...
+    if (diag.length == 1 && diag(0) == 0) {
+      throw new UnsupportedOperationException(
+        "No t values available for this LinearRegressionModel")
+    } else {
+      model.weights.toArray.zip(seCoef).map { x => x._1 / x._2 }
+    }
+  }
+
+  lazy val pVals: Array[Double] = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: `pValues`

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 10]
```diff
@@ -471,6 +484,59 @@ class LinearRegressionSummary private[regression] (
...
 
+  lazy val numInstances: Long = predictions.count()
+
+  lazy val dfe = if (model.getFitIntercept) {
+    numInstances - model.weights.size -1
+  } else {
+    numInstances - model.weights.size
+  }
+
+  lazy val devianceResiduals: Array[Double] = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: It is useful to document that this is weighted.

**Comment by mengxr** [comment]
@yanboliang The implementation looks good to me. I left some comments about comment/documentation. Could you address them today to catch 1.6? It is okay to address the issue with intercept in a follow-up PR. You can create a JIRA for it. Thanks!

File: mllib/src/test/scala/org/apache/spark/ml/regression/LinearRegressionSuite.scala [hunk 11]
```diff
@@ -715,4 +724,63 @@ class LinearRegressionSuite extends SparkFunSuite with MLlibTestSparkContext {
         .sliding(2)
         .forall(x => x(0) >= x(1)))
   }
+
+  test("linear regression training summary with weighted samples by normal solver") {
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/regression/LinearRegressionSuite.scala]: Could you also add a test without intercept?

**Comment by yanboliang** [comment]
@mengxr I created [SPARK-11473](https://issues.apache.org/jira/browse/SPARK-11473) to track the issue of supporting summary statistic for intercept. I can work on it.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 12]
```diff
@@ -474,6 +487,75 @@ class LinearRegressionSummary private[regression] (
...
+          math.pow(label - pred, 2.0) * weight }
+        predictions.select(t(col(model.getPredictionCol), col(model.getLabelCol),
+          col(model.getWeightCol)).as("wse")).agg(sum(col("wse"))).first().getDouble(0)
+      }
+      val sigma2 = rss / degreesOfFreedom
+      diagInvAtWA.map(_ * sigma2).map(math.sqrt(_))
+    }
+  }
+
+  /** T-statistic of estimated coefficients.
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: minor: This is ScalaDoc style. We can fix it in the next update.

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 13]
```diff
@@ -474,6 +487,75 @@ class LinearRegressionSummary private[regression] (
...
+    numInstances - model.coefficients.size - 1
+  } else {
+    numInstances - model.coefficients.size
+  }
+
+  /**
+   * The weighted residuals, the usual residuals rescaled by
+   * the square root of the instance weights.
+   */
+  lazy val devianceResiduals: Array[Double] = {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: I'm late to comment, but am wondering:
* Why do we not return all deviance residuals as a DataFrame?  If we only return min,max, then that should be documented.  But I'd prefer we return a DataFrame with all deviance residuals.
* Should we follow R's example and just call this "residuals"?  That will let us add other types of residuals later (specified via an argument, with a default argument of "deviance").

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 14]
```diff
@@ -474,6 +487,75 @@ class LinearRegressionSummary private[regression] (
...
+      .select(min(col("weightedResiduals")).as("min"), max(col("weightedResiduals")).as("max"))
+      .first()
+    Array(dr.getDouble(0), dr.getDouble(1))
+  }
+
+  /**
+   * Standard error of estimated coefficients.
+   * Note that standard error of estimated intercept is not supported currently.
+   */
+  lazy val coefficientStandardErrors: Array[Double] = {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Should we return a Vector (to match the type of coefficients)?  Same for tValues and pValues.

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/CholeskyDecomposition.scala [hunk 15]
```diff
@@ -40,4 +40,20 @@ private[spark] object CholeskyDecomposition {
...
+   * The input arguments are modified in-place to store the inverse matrix.
+   * @param UAi the upper triangular factor U from the Cholesky factorization A = U**T*U
+   * @param k the dimension of A
+   * @return the upper triangle of the (symmetric) inverse of A
+   */
+  def inverse(UAi: Array[Double], k: Int): Array[Double] = {
+    val info = new intW(0)
+    lapack.dpptri("U", k, UAi, info)
+    val code = info.`val`
+    assert(code == 0, s"lapack.dpptri returned $code.")
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/CholeskyDecomposition.scala]: This throws an AssertionError on failure.  It'd be better to throw a RuntimeError (or one based on the return code, though that may be too much trouble).

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala (hunk 13 above)
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: @jkbradley There is "residuals" already exist, so I call it ```devianceResiduals```. I agree your opinion about adding other types of residuals later, so I think we can try to combine the two functions into one with different arguments. We also need do some code clean up for ```LinearRegressionSummary``` due to redundant arguments, I can finish it in a follow up PR. @mengxr

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Sounds good!

**Pull request closed without merging**